### PR TITLE
Implement floating point strides

### PR DESCRIPTION
**Pull request by xwu**
#### What's in this pull request?

This pull request adds an underscored static method to Strideable and uses this method to specialize the implementation for advancing a value by a given floating point stride distance. New tests are included for the correct function of floating point strides. The changes are in line with suggestions outlined by the core team upon returning SE-0050 for revision, namely:

> [W]e believe the accumulating problem can be fixed without changing Strideable's public API, e.g. by adding underscored protocol requirements for which we can provide specialized implementations using protocol extensions constrained to Stride : FloatingPointType.  We look forward to accepting a pull request to make this change, which doesn't require an evolution proposal, and we expect that change to usefully inform the question of how to change Strideable.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

Add tests for floating point stride

Clarify comment about unreachable endpoint

Add test for floating point stride error accumulation

Restore some formatting

### CODE REVIEW

**Comment by natecook1000** [comment]
This looks great, @xwu! Much more sane output for floating-point strides. For reference:

```swift
let a: [Float] = Array(stride(from: 1.0, through: 2.0, by: 0.1))

// before
// a.count == 10
// a == [1.0, 1.10000002, 1.20000005, 1.30000007, 1.4000001, 1.50000012, 1.60000014, 1.70000017, 1.80000019, 1.90000021]

// after
// a.count == 11
// a == [1.0, 1.10000002, 1.20000005, 1.29999995, 1.39999998, 1.5, 1.60000002, 1.70000005, 1.79999995, 1.9000001, 2.0]

**Comment by lattner** [comment]
@gribozavr or @dabrahams can you take a look at this?

**Comment by gribozavr** [comment]
It is in my queue.

File: stdlib/public/core/Stride.swift.gyb [hunk 1]
```diff
@@ -125,22 +131,54 @@ public func -= <T : UnsignedInteger>(
...
+  }
+}
+
+/// An iterator for `StrideTo<Element>`.
 public struct StrideToIterator<Element : Strideable> : IteratorProtocol {
-  internal var _current: Element
+  internal let _start: Element
   internal let _end: Element
   internal let _stride: Element.Stride
+  internal var _current: (index: Int?, value: Element)
```
Comment (dabrahams) [inline stdlib/public/core/Stride.swift.gyb]: It’s a nit, but in general it would be better to avoid reordering properties, to minimize diffs.

**Comment by dabrahams** [comment]
LGTM!

**Comment by dabrahams** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
LGTM too.

**Comment by dabrahams** [comment]
THANK YOU!

**Comment by xwu** [comment]
Thank you all!

**Pull request merged**